### DISCUSSION TITLE
About getServerSideProps not working in next.js 13.1.6.

### DISCUSSION THREAD

**Discussion by NicoHasNoBugs** [vercel/next.js#47893] [Help] [answered]
### Summary

In the next.js 13.1.6 version, after packaging and publishing to the server, the getServerSideProps method is not executed.It works fine locally.
We use two PM2 servers.

### Additional information

```js
export async function getServerSideProps (context) {
    const { req, locale, res } = context || {};
    const { query = {} } = req;
    
    /* When this line is not printed, the page cannot accept props */
    console.log(req, locale, res);
    
    res.setHeader(
        'Cache-Control',
        'no-cache, no-store, max-age=0, must-revalidate'
    );
    
    /* Get some interface data */
    const pageData = await getDataHandle({ locale, query });
    
    return {
        props: {
            pageData,
            locale,
            query,
        },
    }
}

function Page (props) {
    useEffect(() => {
        setTimeout(() => {
            /* some events */
        }, 8000);
        
        getData();
        
    }, []);

    const getData = async () => {
        /* Unable to get locale, query, pageData, it is undefined */
        const { locale, query, pageData } = props;
        
    }
}



export default Page;
```

**Comment by icyJoseph** [comment]
Now this is really weird. Do you see requests leaving the browser for `/_next/data` that end with .json, for the page you are navigating to?

How does navigating to the page directly look like?  Does it work then?

**Reply by NicoHasNoBugs** [reply]
Thanks for answering! 
We directly open this page getServerSideProps can run normally.
But choosing to disable cookies when the website prompts whether to use cookies, and jumping to this page from other pages will make this problem inevitable.

**Reply by icyJoseph** [reply]
Aha, alright that's a start.

- Do you have a Service Worker?
- Do you see requests for `_next/data` leaving the browser when you navigate to this page, from other pages in your app?

**Reply by NicoHasNoBugs** [reply]
-  Yes
-  I can see that this page is requested, and if I print something in getData, it can be printed.

**Reply by icyJoseph** [reply]
Often Service Worker frameworks use the Stale-While-Revalidate strategy by default, which defeats the purpose of GetServerSideProps. You want to use network-first-with-fallback-on-cache:

- Not good for GSSP: https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#stale-while-revalidate
- Better: https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#network-first-network-falling-back-to-cache

Are you using next-pwa? If so, do this: https://github.com/vercel/next.js/discussions/42405#discussioncomment-4046461

```js
// next.config.js

/** @type {import('next').NextConfig} */
const runtimeCaching = require("next-pwa/cache");

const nextDataIndex = runtimeCaching.findIndex(
  (entry) => entry.options.cacheName === "next-data"
);

if (nextDataIndex !== -1) {
  runtimeCaching[nextDataIndex].handler = "NetworkFirst";
} else {
  throw new Error("Failed to find next-data object in runtime caching");
}

const withPWA = require("next-pwa")({
  dest: "public",
  runtimeCaching,
  register: true,
  skipWaiting: true,
  buildExcludes: [/middleware-manifest.json$/],
  disable: process.env.NODE_ENV === 'development'
})

module.exports = withPWA({
  reactStrictMode: false
});
```

With Stale-While-Revalidate, the call is intercepted by the Service Worker, and it never hits the server, until you request it again.

**Reply by NicoHasNoBugs** [reply, marked as answer]
Thank you!
The components you mentioned were not used in the project, but the knowledge you mentioned gave me inspiration.
The previous page used router.prefetch to preload pages with SSR. I removed the router.prefetch, it can already run normally.

**Reply by icyJoseph** [reply]
Aha! Perfect! I think you can close this issue then :) Glad to have helped!

**Reply by NicoHasNoBugs** [reply]
Although this fixed the problem, I haven't figured out what the rationale is for this to fix the problem? Can I ask you by the way? 
Thank you!
